Reorganized Plutus user guide on the Docusaurus platform initial deployment (#6177)
IntersectMBO/plutus: docusaurus/docs/adr/adr2.md

  userComputation state
  ...
```

### Argument: coroutine system

Why a coroutine system? 
In short, structuring the code this way will ease our future work. 
Some of the advantages are mentioned above already. 
Here is a summary:

- The debugger is naturally a coroutine, where one routine is the user and the other is the CEK machine, and they take turns to suspend and pass data and control to each other in a debugging session. The literature has contributed a good way to design/implement a coroutine. It makes sense to implement a well studied design.
- We can probably reuse the same monad (`CekM`) in the steppable CEK machine, because we can convert it to/from `IO` via `unsafeSTToIO` and `unsafeIOToST`.
- It should be easier when we add more functionalities because multiple suspension functors and computations can be composed using [coproducts](https://www.cambridge.org/core/services/aop-cambridge-core/content/view/14416CB20C4637164EA9F77097909409/S0956796808006758a.pdf/data-types-a-la-carte.pdf).
- This should also play nicely when we implement Debug Adapter Protocol for the debugger later on.

## Implications

> In summary, we proposed to implement the debugging machine as a coroutine system with "steps". 
> This implies that:
> 
> - We have to maintain the CEK machine. E.g., we need to check its conformance.
> - We will add a debugger for our users. We can give users more information at each evaluation step.
> - We will need to write some tests to ensure that the debugging machine continuously output reasonable information.

Reviewer: ```suggestion
In summary, we proposed to implement the debugging machine as a coroutine system with 'steps'. 
This implies that:

- We have to maintain the CEK machine (eg, we need to check its conformance)
- We will add a debugger for our users providing them with more information at each evaluation step
- We will need to write some tests to ensure that the debugging machine continuously outputs reasonable information.
```
